client/my-sites/checklist/wpcom-checklist/index.jsx

```diff
@@ -0,0 +1,487 @@
+import { requestSiteChecklistTaskUpdate } from 'state/checklist/actions';
+import { getCurrentUser, isCurrentUserEmailVerified } from 'state/current-user/selectors';
+import userFactory from 'lib/user';
+
+const userLib = userFactory();
+
+const query = { type: 'any', number: 10, order_by: 'ID', order: 'ASC' };
+
+class WpcomChecklist extends PureComponent {
+	static propTypes = {
+		createNotice: PropTypes.func.isRequired,
+		designType: PropTypes.oneOf( [ 'blog', 'page', 'portfolio' ] ),
+		recordTracksEvent: PropTypes.func.isRequired,
+		requestGuidedTour: PropTypes.func.isRequired,
+		requestSiteChecklistTaskUpdate: PropTypes.func.isRequired,
+		siteId: PropTypes.number,
+		siteSlug: PropTypes.string,
+		taskStatuses: PropTypes.object,
+		viewMode: PropTypes.oneOf( [ 'checklist', 'banner', 'navigation', 'notification' ] ),
+	};
+
+	static defaultProps = {
+		viewMode: 'checklist',
+	};
+
+	state = {
+		pendingRequest: false,
+		emailSent: false,
+		error: null,
+	};
+
+	canShow = () => {
+		if ( ! this.props.isEligibleForDotcomChecklist || ! this.props.isSiteSection ) {
+			return false;
+		}
+
+		return true;
+	};
+
+	handleTaskStart = ( { task, tourId, url } ) => () => {
+		if ( ! tourId && ! url ) {
+			return;
+		}
+
+		const location = 'banner' === this.props.viewMode ? 'checklist_banner' : 'checklist_show';
+
+		this.props.recordTracksEvent( 'calypso_checklist_task_start', {
+			checklist_name: 'new_blog',
+			location,
+			step_name: task.id,
+		} );
+
+		if ( tourId && ! task.isCompleted && isDesktop() ) {
+			this.props.requestGuidedTour( tourId );
+		}
+
+		if ( url ) {
+			page.show( url );
+		}
+	};
+
+	handleTaskDismiss = taskId => () => {
+		const { siteId } = this.props;
+
+		if ( taskId ) {
+			this.props.createNotice( 'is-success', 'You completed a task!' );
+			this.props.requestSiteChecklistTaskUpdate( siteId, taskId );
+		}
+	};
+
+	handleSendVerificationEmail = e => {
+		e.preventDefault();
+
+		if ( this.state.pendingRequest ) {
+			return;
+		}
+
+		this.setState( {
+			pendingRequest: true,
+		} );
+
+		userLib.sendVerificationEmail( ( error, response ) => {
+			this.setState( {
+				emailSent: response && response.success,
+				error: error,
+				pendingRequest: false,
+			} );
+		} );
+	};
+
+	verificationTaskButtonText() {
+		const { translate } = this.props;
+		if ( this.state.pendingRequest ) {
+			return translate( 'Sending…' );
+		}
+
+		if ( this.state.error ) {
+			return translate( 'Error' );
+		}
+
+		if ( this.state.emailSent ) {
+			return translate( 'Email sent' );
+		}
+
+		return translate( 'Resend email' );
+	}
+
+	render() {
+		const {
+			designType,
+			siteId,
+			taskStatuses,
+			viewMode,
+			updateCompletion,
+			setNotification,
+			setStoredTask,
+			closePopover,
+			showNotification,
+			storedTask,
+		} = this.props;
+
+		const canShowChecklist = this.canShow();
+		const taskList = new WpcomTaskList( taskStatuses, { designType } );
```

Comment:
Here we create a new instance on every rerender. Doesn't it cause some unwanted rerenders?

Reply:
We're already inside the render method, so the component and its children are already going to render. That said, this could make a difference if one of the children is looking at its props to determine whether or not to re-render itself (e.g. by being a pure component, or by having a `shouldComponentUpdate`).

In order to cover those situations, I added memoization to this.